File: apps/cli/internal/daemon/pr_tracker.go
package daemon

import (
	"context"
	"fmt"
	"strings"
	"sync"
	"time"

	"yishan/apps/cli/internal/api"
	cliruntime "yishan/apps/cli/internal/runtime"
	"yishan/apps/cli/internal/workspace"

	"github.com/rs/zerolog/log"
)

const workspacePullRequestPollInterval = 5 * time.Minute

const ghUnknownGitHubHostErrorFragment = "none of the git remotes configured for this repository point to a known github host"

type workspacePRTracker struct {
	mu      sync.Mutex
	manager *workspace.Manager
	// active maps workspaceID → Workspace for all workspaces currently being
	// tracked. Storing the full Workspace avoids calling manager.List() on
	// every poll tick and filtering by active map membership.
	active          map[string]workspace.Workspace
	inFlight        map[string]bool
	started         bool
	done            chan struct{}
	publish         func(frontendEvent)
	inspectResolver func(context.Context, string) (workspace.GitInspectResult, error)
	branchResolver  func(context.Context, string) (string, error)
	prResolver      func(context.Context, string, string) (workspace.GitBranchPullRequestStatus, error)
	detailResolver  func(context.Context, string, string) (workspace.GitBranchPullRequestStatus, error)
}

func newWorkspacePRTracker(manager *workspace.Manager, publish func(frontendEvent)) *workspacePRTracker {
	tracker := &workspacePRTracker{
		manager:  manager,
		active:   make(map[string]workspace.Workspace),
		inFlight: make(map[string]bool),
		done:     make(chan struct{}),
		publish:  publish,
	}
	tracker.branchResolver = func(ctx context.Context, root string) (string, error) {
		ws, ok := manager.FindWorkspaceByPath(root)
		if !ok {
			return "", workspace.NewRPCError(rpcCodeNotFound, "workspace not found")
		}
		return manager.GitCurrentBranch(ctx, ws.ID)
	}
	tracker.inspectResolver = func(ctx context.Context, root string) (workspace.GitInspectResult, error) {
		return manager.GitInspect(ctx, root)
	}
	tracker.prResolver = func(ctx context.Context, root string, branch string) (workspace.GitBranchPullRequestStatus, error) {
		ws, ok := manager.FindWorkspaceByPath(root)
		if !ok {
			return workspace.GitBranchPullRequestStatus{}, workspace.NewRPCError(rpcCodeNotFound, "workspace not found")
		}
		return manager.GitBranchPullRequestLite(ctx, ws.ID, branch)
	}
	tracker.detailResolver = func(ctx context.Context, root string, branch string) (workspace.GitBranchPullRequestStatus, error) {
		ws, ok := manager.FindWorkspaceByPath(root)
		if !ok {
			return workspace.GitBranchPullRequestStatus{}, workspace.NewRPCError(rpcCodeNotFound, "workspace not found")
		}
		return manager.GitBranchPullRequestWithDetails(ctx, ws.ID, branch)
	}
	return tracker
}

func (t *workspacePRTracker) EnsureTracked(worktreePath string, refreshImmediately bool) {
	if strings.TrimSpace(worktreePath) == "" {
		return
	}

	ws, ok := t.manager.FindWorkspaceByPath(worktreePath)
	if !ok {
		return
	}

	if !t.shouldTrackWorkspacePullRequest(ws) {
		t.setWorkspacePullRequest(ws, nil, false)
		return
	}

	t.mu.Lock()
	if !t.started {
		t.started = true
		go t.pollLoop()
	}
	t.active[ws.ID] = ws
	t.mu.Unlock()

	if refreshImmediately {
		go t.RefreshWorkspaceByPath(worktreePath)
	}
}

func (t *workspacePRTracker) StopTracking(workspaceID string) {
	t.mu.Lock()
	defer t.mu.Unlock()
	delete(t.active, workspaceID)
}

// Stop shuts down the background poll loop. It is safe to call multiple times.
func (t *workspacePRTracker) Stop() {
	t.mu.Lock()
	defer t.mu.Unlock()
	select {
	case <-t.done:
		// already closed
	default:
		close(t.done)
	}
}

func (t *workspacePRTracker) RefreshWorkspaceByPath(worktreePath string) {
	ws, ok := t.manager.FindWorkspaceByPath(worktreePath)
	if !ok {
		log.Warn().Str("path", worktreePath).Msg("workspace PR refresh skipped because workspace path is not open")
		return
	}

	t.mu.Lock()
	_, tracked := t.active[ws.ID]
	t.mu.Unlock()
	if !tracked {
		log.Debug().Str("workspaceId", ws.ID).Str("path", ws.Path).Msg("workspace PR refresh skipped because workspace is no longer active")
		return
	}

	if !t.beginRefresh(ws.ID) {
		log.Debug().Str("workspaceId", ws.ID).Str("path", ws.Path).Msg("workspace PR refresh skipped because another refresh is already running")
		return
	}
	defer t.endRefresh(ws.ID)
	if err := t.refreshWorkspace(ws); err != nil {
		log.Debug().Err(err).Str("workspaceId", ws.ID).Str("path", ws.Path).Msg("failed to refresh workspace pull request state")
	}
}

func (t *workspacePRTracker) pollLoop() {
	ticker := time.NewTicker(workspacePullRequestPollInterval)
	defer ticker.Stop()

	for {
		select {
		case <-t.done:
			log.Debug().Msg("workspace PR tracker poll loop stopped")
			return
		case <-ticker.C:
		}

		t.mu.Lock()
		// Snapshot the tracked workspaces under the lock, then release before
		// making network calls. This avoids holding mu during the gh CLI calls.
		tracked := make([]workspace.Workspace, 0, len(t.active))
		for _, ws := range t.active {
			tracked = append(tracked, ws)
		}
		t.mu.Unlock()

		for _, ws := range tracked {
			if !t.beginRefresh(ws.ID) {
				continue
			}

			if err := t.refreshWorkspace(ws); err != nil {
				log.Debug().Err(err).Str("workspaceId", ws.ID).Str("path", ws.Path).Msg("failed to poll workspace pull request state")
			}
			t.endRefresh(ws.ID)
		}
	}
}

func (t *workspacePRTracker) beginRefresh(workspaceID string) bool {
	t.mu.Lock()
	defer t.mu.Unlock()
	if t.inFlight[workspaceID] {
		return false
	}
	t.inFlight[workspaceID] = true
	return true
}

func (t *workspacePRTracker) endRefresh(workspaceID string) {
	t.mu.Lock()
	defer t.mu.Unlock()
	delete(t.inFlight, workspaceID)
}

func (t *workspacePRTracker) refreshWorkspace(ws workspace.Workspace) error {
	ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
	defer cancel()

	branch, err := t.branchResolver(ctx, ws.Path)
	if err != nil {
		if shouldDisableTrackingForBranchError(err) {
			t.setWorkspacePullRequest(ws, nil, false)
			log.Debug().Err(err).Str("workspaceId", ws.ID).Str("path", ws.Path).Msg("workspace PR refresh disabled tracking because branch could not be resolved")
			return nil
		}
		log.Warn().Err(err).Str("workspaceId", ws.ID).Str("path", ws.Path).Msg("workspace PR refresh failed to resolve branch")
		return err
	}
	branch = strings.TrimSpace(branch)
	log.Debug().Str("workspaceId", ws.ID).Str("path", ws.Path).Str("branch", branch).Msg("workspace PR refresh resolved branch")
	if branch == "" || branch == "HEAD" {
		t.setWorkspacePullRequest(ws, nil, true)
		log.Debug().Str("workspaceId", ws.ID).Str("path", ws.Path).Msg("workspace PR refresh cleared PR because branch is empty or detached")
		return nil
	}

	pr, err := t.detailResolver(ctx, ws.Path, branch)
	if err != nil {
		if shouldErrDisableTracking(err) {
			t.setWorkspacePullRequest(ws, nil, false)
			log.Debug().Err(err).Str("workspaceId", ws.ID).Str("path", ws.Path).Str("branch", branch).Msg("workspace PR refresh disabled tracking for repository without PR support")
			return nil
		}
		log.Warn().Err(err).Str("workspaceId", ws.ID).Str("path", ws.Path).Str("branch", branch).Msg("workspace PR refresh failed to resolve pull request")
		return err
	}
	if !pr.Found {
		t.setWorkspacePullRequest(ws, nil, true)
		log.Debug().Str("workspaceId", ws.ID).Str("path", ws.Path).Str("branch", branch).Msg("workspace PR refresh found no pull request")
		return nil
	}

	status := normalizeWorkspacePullRequestStatus(pr)
	bound := &workspace.WorkspacePullRequest{
		Number:         pr.Number,
		Title:          pr.Title,
		URL:            pr.URL,
		Branch:         pr.HeadRefName,
		BaseBranch:     pr.BaseRefName,
		GitHubState:    pr.State,
		Status:         status,
		ReviewDecision: pr.ReviewDecision,
		IsDraft:        pr.IsDraft,
		Complete:       status == "merged",
		UpdatedAt:      nowRFC3339Nano(),
		Checks:         pr.Checks,
		Deployments:    pr.Deployments,
	}
	complete := status == "merged"
	t.setWorkspacePullRequest(ws, bound, !complete)
	log.Debug().
		Str("workspaceId", ws.ID).
		Str("path", ws.Path).
		Str("branch", branch).
		Int("pullRequestNumber", pr.Number).
		Str("pullRequestStatus", status).
		Bool("complete", complete).
		Msg("workspace PR refresh synced pull request")
	return nil
}

func shouldErrDisableTracking(err error) bool {
	if err == nil {
		return false
	}
	message := strings.ToLower(strings.TrimSpace(err.Error()))
	return strings.Contains(message, ghUnknownGitHubHostErrorFragment) ||
		strings.Contains(message, "no git remote") ||
		strings.Contains(message, "no remotes")
}

func shouldDisableTrackingForBranchError(err error) bool {
	if err == nil {
		return false
	}
	message := strings.ToLower(strings.TrimSpace(err.Error()))
	return strings.Contains(message, "workspace is not on a branch") ||
		strings.Contains(message, "ambiguous argument 'head'") ||
		strings.Contains(message, "unknown revision or path not in the working tree")
}

func (t *workspacePRTracker) setWorkspacePullRequest(ws workspace.Workspace, pr *workspace.WorkspacePullRequest, keepActive bool) {
	previousPullRequest := ws.PullRequest
	if err := t.manager.SetWorkspacePullRequest(ws.ID, pr); err != nil {
		return
	}
	if prMeaningfullyChanged(previousPullRequest, pr) {
		if t.publish != nil {
			t.publish(frontendEvent{
				Topic: "workspacePullRequestUpdated",
				Payload: map[string]any{
					"workspaceId":           ws.ID,
					"workspaceWorktreePath": ws.Path,
					"pullRequest":           pr,
				},
			})
		}
	}

	t.mu.Lock()
	defer t.mu.Unlock()
	if keepActive {
		if _, ok := t.active[ws.ID]; ok {
			ws.PullRequest = pr
			t.active[ws.ID] = ws
		}
	} else {
		delete(t.active, ws.ID)
	}

	// Persist to api-service only when meaningful PR fields changed (excluding
	// UpdatedAt which is set on every refresh and would always differ).
	if pr != nil && prMeaningfullyChanged(previousPullRequest, pr) {
		go t.persistPullRequest(ws.ID, pr)
	}
}

// prMeaningfullyChanged returns true when the PR fields that matter for
// persistence have changed, ignoring UpdatedAt which is always refreshed.
func prMeaningfullyChanged(prev, next *workspace.WorkspacePullRequest) bool {
	if prev == nil && next == nil {
		return false
	}
	if prev == nil || next == nil {
		return true
	}
	return prev.Number != next.Number ||
		prev.Title != next.Title ||
		prev.URL != next.URL ||
		prev.Branch != next.Branch ||
		prev.BaseBranch != next.BaseBranch ||
		prev.GitHubState != next.GitHubState ||
		prev.Status != next.Status ||
		prev.ReviewDecision != next.ReviewDecision ||
		prev.IsDraft != next.IsDraft ||
		prev.Complete != next.Complete ||
		!checksEqual(prev.Checks, next.Checks) ||
		!deploymentsEqual(prev.Deployments, next.Deployments)
}

// checksEqual compares two check slices field-by-field.
// Length-first comparison short-circuits the common case of different counts.
func checksEqual(a, b []workspace.GitPullRequestCheck) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i].Name != b[i].Name ||
			a[i].State != b[i].State ||
			a[i].Workflow != b[i].Workflow ||
			a[i].Description != b[i].Description ||
			a[i].URL != b[i].URL {
			return false
		}
	}
	return true
}

// deploymentsEqual compares two deployment slices field-by-field.
// Length-first comparison short-circuits the common case of different counts.
func deploymentsEqual(a, b []workspace.GitPullRequestDeployment) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i].ID != b[i].ID ||
			a[i].Environment != b[i].Environment ||
			a[i].State != b[i].State ||
			a[i].Description != b[i].Description ||
			a[i].EnvironmentURL != b[i].EnvironmentURL {
			return false
		}
	}
	return true
}

func normalizeWorkspacePullRequestStatus(pr workspace.GitBranchPullRequestStatus) string {
	state := strings.ToUpper(strings.TrimSpace(pr.State))
	if state == "MERGED" || strings.TrimSpace(pr.MergedAt) != "" {
		return "merged"
	}
	if pr.IsDraft {
		return "draft"
	}
	if strings.EqualFold(strings.TrimSpace(pr.ReviewDecision), "REVIEW_REQUIRED") {
		return "review"
	}
	if state == "OPEN" {
		return "open"
	}
	if state == "CLOSED" {
		return "closed"
	}
	return strings.ToLower(state)
}

// persistPullRequest writes a PR snapshot to the api-service.
// Called in a goroutine; failures are logged and do not affect local state.
func (t *workspacePRTracker) persistPullRequest(workspaceID string, pr *workspace.WorkspacePullRequest) {
	if !cliruntime.APIConfigured() {
		return
	}

	ws, err := t.manager.GetWorkspace(workspaceID)
	if err != nil {
		log.Warn().Err(err).Str("workspaceId", workspaceID).Msg("pr persist: workspace not found")
		return
	}
	if ws.OrgID == "" || ws.ProjectID == "" {
		log.Debug().Str("workspaceId", workspaceID).Msg("pr persist: skipped — orgId or projectId not set on workspace")
		return
	}

	// Map daemon status to api-service state.
	state := pr.Status
	if state == "draft" || state == "review" {
		state = "open"
	}
	if state != "open" && state != "closed" && state != "merged" {
		state = "open"
	}

	resolvedAt := ""
	if state == "merged" || state == "closed" {
		resolvedAt = nowRFC3339Nano()
	}

	metadata := map[string]any{
		"isDraft":        pr.IsDraft,
		"reviewDecision": pr.ReviewDecision,
	}
	if len(pr.Checks) > 0 {
		checks := make([]map[string]any, 0, len(pr.Checks))
		for _, c := range pr.Checks {
			checks = append(checks, map[string]any{
				"name":        c.Name,
				"workflow":    c.Workflow,
				"state":       c.State,
				"description": c.Description,
				"url":         c.URL,
			})
		}
		metadata["checks"] = checks
	}
	if len(pr.Deployments) > 0 {
		deployments := make([]map[string]any, 0, len(pr.Deployments))
		for _, d := range pr.Deployments {
			deployments = append(deployments, map[string]any{
				"id":             d.ID,
				"environment":    d.Environment,
				"state":          d.State,
				"description":    d.Description,
				"environmentUrl": d.EnvironmentURL,
			})
		}
		metadata["deployments"] = deployments
	}

	input := api.UpsertWorkspacePullRequestInput{
		PrID:       fmt.Sprintf("%d", pr.Number),
		Title:      pr.Title,
		URL:        pr.URL,
		Branch:     pr.Branch,
		BaseBranch: pr.BaseBranch,
		State:      state,
		Metadata:   metadata,
		DetectedAt: pr.UpdatedAt,
		ResolvedAt: resolvedAt,
	}

	if _, err := cliruntime.APIClient().UpsertWorkspacePullRequest(ws.OrgID, ws.ProjectID, workspaceID, input); err != nil {
		log.Warn().Err(err).Str("workspaceId", workspaceID).Str("prId", input.PrID).Str("state", state).Msg("pr persist: failed to upsert to api-service")
		return
	}
	log.Debug().Str("workspaceId", workspaceID).Str("prId", input.PrID).Str("state", state).Msg("pr persist: upserted to api-service")
}

File: apps/cli-rs/src/daemon/pr_tracker.rs
#![allow(dead_code)]

use crate::daemon::event_hub::{EventHub, FrontendEvent};
use crate::workspace::manager::WorkspaceManager;
use crate::workspace::types::{Workspace, WorkspacePullRequest};
use serde_json::json;
use std::collections::{HashMap, HashSet};
use std::sync::{Arc, Mutex};
use std::time::Duration;
use tokio::sync::Notify;
use tracing::{debug, warn};

const POLL_INTERVAL: Duration = Duration::from_secs(5 * 60); // 5 minutes
const GH_UNKNOWN_HOST_FRAGMENT: &str =
    "none of the git remotes configured for this repository point to a known github host";

struct Inner {
    active: HashMap<String, Workspace>,
    in_flight: HashSet<String>,
}

/// Background poll tracker for workspace pull-request state.
/// Publishes `workspacePullRequestUpdated` events when meaningful PR fields change.
pub struct PrTracker {
    inner: Arc<Mutex<Inner>>,
    cancel: Arc<Notify>,
    manager: Arc<WorkspaceManager>,
    events: Arc<EventHub>,
}

impl PrTracker {
    pub fn new(manager: Arc<WorkspaceManager>, events: Arc<EventHub>) -> Arc<Self> {
        Arc::new(Self {
            inner: Arc::new(Mutex::new(Inner {
                active: HashMap::new(),
                in_flight: HashSet::new(),
            })),
            cancel: Arc::new(Notify::new()),
            manager,
            events,
        })
    }

    /// Start the background poll loop. Safe to call once.
    pub fn start(self: &Arc<Self>) {
        let this = self.clone();
        tokio::spawn(async move { this.poll_loop().await });
    }

    /// Register a workspace for PR tracking; optionally trigger an immediate refresh.
    pub fn ensure_tracked(self: &Arc<Self>, worktree_path: &str, refresh_immediately: bool) {
        let path = worktree_path.trim();
        if path.is_empty() {
            return;
        }
        let ws = match self.find_by_path(path) {
            Some(w) => w,
            None => return,
        };
        {
            let mut guard = self.inner.lock().unwrap();
            guard.active.insert(ws.id.clone(), ws.clone());
        }
        if refresh_immediately {
            let this = self.clone();
            let ws2 = ws.clone();
            tokio::spawn(async move { this.do_refresh(ws2).await });
        }
    }

    /// Remove a workspace from tracking.
    pub fn stop_tracking(&self, workspace_id: &str) {
        self.inner.lock().unwrap().active.remove(workspace_id);
    }

    /// Signal the poll loop to stop.
    pub fn stop(&self) {
        self.cancel.notify_waiters();
    }

    /// Immediately refresh PR state for the workspace at the given path.
    pub async fn refresh_by_path(self: &Arc<Self>, worktree_path: &str) {
        let ws = match self.find_by_path(worktree_path) {
            Some(w) => w,
            None => {
                warn!(path = worktree_path, "PR refresh: workspace not found");
                return;
            }
        };
        let tracked = self.inner.lock().unwrap().active.contains_key(&ws.id);
        if !tracked {
            debug!(id = ws.id, "PR refresh skipped: not tracked");
            return;
        }
        if !self.begin_refresh(&ws.id) {
            debug!(id = ws.id, "PR refresh skipped: already in flight");
            return;
        }
        self.do_refresh(ws.clone()).await;
        self.end_refresh(&ws.id);
    }

    // ── internals ─────────────────────────────────────────────────────────────

    async fn poll_loop(self: &Arc<Self>) {
        let mut interval = tokio::time::interval(POLL_INTERVAL);
        interval.tick().await; // skip first immediate tick
        loop {
            tokio::select! {
                _ = self.cancel.notified() => {
                    debug!("PR tracker poll loop stopped");
                    return;
                }
                _ = interval.tick() => {
                    let tracked: Vec<Workspace> = {
                        self.inner.lock().unwrap().active.values().cloned().collect()
                    };
                    for ws in tracked {
                        if !self.begin_refresh(&ws.id) { continue; }
                        let this = self.clone();
                        tokio::spawn(async move {
                            this.do_refresh(ws.clone()).await;
                            this.end_refresh(&ws.id);
                        });
                    }
                }
            }
        }
    }

    fn begin_refresh(&self, workspace_id: &str) -> bool {
        let mut guard = self.inner.lock().unwrap();
        if guard.in_flight.contains(workspace_id) {
            return false;
        }
        guard.in_flight.insert(workspace_id.to_string());
        true
    }

    fn end_refresh(&self, workspace_id: &str) {
        self.inner.lock().unwrap().in_flight.remove(workspace_id);
    }

    async fn do_refresh(self: &Arc<Self>, ws: Workspace) {
        // All git operations are blocking (subprocess calls) — run on blocking thread.
        let mgr = self.manager.clone();
        let events = self.events.clone();
        let inner = self.inner.clone();
        let ws_id = ws.id.clone();
        let _ws_path = ws.path.clone();
        let ws2 = ws.clone();

        tokio::task::spawn_blocking(move || {
            refresh_workspace_sync(&mgr, &events, &inner, ws2);
        })
        .await
        .unwrap_or_else(|e| debug!(err = %e, id = ws_id, "PR refresh task panicked"));
    }

    fn find_by_path(&self, path: &str) -> Option<Workspace> {
        let path = path.trim();
        self.manager.list().into_iter().find(|w| w.path.trim() == path)
    }
}

fn refresh_workspace_sync(
    manager: &WorkspaceManager,
    events: &EventHub,
    inner: &Mutex<Inner>,
    ws: Workspace,
) {
    // Resolve current branch (sync subprocess).
    let branch = match manager.gits.current_branch(&ws.path) {
        Ok(b) => b.trim().to_string(),
        Err(e) => {
            let msg = e.to_string().to_lowercase();
            if should_disable_for_branch_error(&msg) {
                set_pull_request(manager, events, inner, &ws, None, false);
                return;
            }
            debug!(err = %e, id = ws.id, "PR refresh: branch resolution failed");
            return;
        }
    };

    if branch.is_empty() || branch == "HEAD" {
        set_pull_request(manager, events, inner, &ws, None, true);
        return;
    }

    // Fetch PR details.
    let pr_result = manager.gits.branch_pull_request(&ws.path, &branch);
    let pr_status = match pr_result {
        Ok(s) => s,
        Err(e) => {
            let msg = e.to_string().to_lowercase();
            if should_disable_for_pr_error(&msg) {
                set_pull_request(manager, events, inner, &ws, None, false);
                return;
            }
            debug!(err = %e, id = ws.id, "PR refresh: pr resolution failed");
            return;
        }
    };

    if !pr_status.found {
        set_pull_request(manager, events, inner, &ws, None, true);
        return;
    }

    let status = normalize_pr_status(&pr_status);
    let complete = status == "merged";
    let pr = WorkspacePullRequest {
        number: pr_status.number,
        title: pr_status.title.clone(),
        url: pr_status.url.clone(),
        branch: pr_status.head_ref_name.clone(),
        base_branch: pr_status.base_ref_name.clone(),
        github_state: pr_status.state.clone(),
        status: status.clone(),
        review_decision: pr_status.review_decision.clone(),
        is_draft: pr_status.is_draft,
        complete,
        updated_at: chrono::Utc::now().to_rfc3339_opts(chrono::SecondsFormat::Nanos, true),
        checks: pr_status.checks.clone(),
        deployments: pr_status.deployments.clone(),
    };

    set_pull_request(manager, events, inner, &ws, Some(pr), !complete);
}

fn set_pull_request(
    manager: &WorkspaceManager,
    events: &EventHub,
    inner: &Mutex<Inner>,
    ws: &Workspace,
    pr: Option<WorkspacePullRequest>,
    keep_active: bool,
) {
    // Get previous value.
    let prev = manager.get(&ws.id).ok().and_then(|w| w.pull_request);
    let _ = manager.set_pull_request(&ws.id, pr.clone());

    if pr_meaningfully_changed(prev.as_ref(), pr.as_ref()) {
        events.publish(FrontendEvent::new(
            "workspacePullRequestUpdated",
            json!({
                "workspaceId": ws.id,
                "workspaceWorktreePath": ws.path,
                "pullRequest": pr,
            }),
        ));
    }

    let mut guard = inner.lock().unwrap();
    if keep_active {
        if let Some(entry) = guard.active.get_mut(&ws.id) {
            entry.pull_request = pr;
        }
    } else {
        guard.active.remove(&ws.id);
    }
}

// ── helpers ────────────────────────────────────────────────────────────────

fn normalize_pr_status(pr: &crate::workspace::types::GitBranchPullRequestStatus) -> String {
    let state = pr.state.trim().to_uppercase();
    if state == "MERGED" || !pr.merged_at.trim().is_empty() {
        return "merged".into();
    }
    if pr.is_draft {
        return "draft".into();
    }
    if pr.review_decision.trim().eq_ignore_ascii_case("REVIEW_REQUIRED") {
        return "review".into();
    }
    match state.as_str() {
        "OPEN" => "open".into(),
        "CLOSED" => "closed".into(),
        _ => state.to_lowercase(),
    }
}

fn pr_meaningfully_changed(
    prev: Option<&WorkspacePullRequest>,
    next: Option<&WorkspacePullRequest>,
) -> bool {
    match (prev, next) {
        (None, None) => false,
        (None, Some(_)) | (Some(_), None) => true,
        (Some(a), Some(b)) => {
            a.number != b.number
                || a.title != b.title
                || a.url != b.url
                || a.branch != b.branch
                || a.base_branch != b.base_branch
                || a.github_state != b.github_state
                || a.status != b.status
                || a.review_decision != b.review_decision
                || a.is_draft != b.is_draft
                || a.complete != b.complete
        }
    }
}

fn should_disable_for_branch_error(msg: &str) -> bool {
    msg.contains("workspace is not on a branch")
        || msg.contains("ambiguous argument 'head'")
        || msg.contains("unknown revision or path not in the working tree")
}

fn should_disable_for_pr_error(msg: &str) -> bool {
    msg.contains(GH_UNKNOWN_HOST_FRAGMENT)
        || msg.contains("no git remote")
        || msg.contains("no remotes")
}
